### PR TITLE
C# 7 refactor + Logging

### PR DESCRIPTION
**Pull request by zeroZshadow**
#### Description
Refactor code to use the new C# 7 functionality unity provides
Small refactor on Logging to allow de-duplication of logger name everywhere.

#### Primary reviewers
@jamiebrynes7 @martinzlocha

### CODE REVIEW

File: workers/unity/Packages/com.improbable.gdk.core/Logging/Logger.cs [hunk 1]
```diff
@@ -0,0 +1,21 @@
+using UnityEngine;
+
+namespace Improbable.Gdk.Core
+{
+    public class Logger : ILogger
+    {
+        private readonly ILogDispatcher dispatcher;
```
Comment (samiwh) [inline workers/unity/Packages/com.improbable.gdk.core/Logging/Logger.cs]: We are we adding our a wrapper around our own logger and using another similarly named interface?

File: workers/unity/Packages/com.improbable.gdk.core/Systems/CleanReactiveComponentsSystem.cs [hunk 2]
```diff
@@ -33,13 +33,14 @@ private void GenerateComponentGroups()
...
+                foreach (var componentType in translationUnit.CleanUpComponentTypes)
                 {
-                    translationUnit.CleanUpComponentGroups.Add(GetComponentGroup(componentType));
+                    var componentGroup = GetComponentGroup(componentType);
+                    translationUnit.CleanUpComponentGroups.Add(componentGroup);
                 }
             }
 
-            MethodInfo addRemoveComponentActionMethod =
+            var addRemoveComponentActionMethod =
```
[the remark this answers is not in the document]
Comment (zeroZshadow) [reply workers/unity/Packages/com.improbable.gdk.core/Systems/CleanReactiveComponentsSystem.cs]: If it was a value that was used throughout a longer scope, I can see why you might not want to use `var`. But with a short scope like this, I see no reason for it to be explicitly typed?

**Comment by zeroZshadow** [comment]
This refactor doesn't have to be merged in. Its something I did on Friday to enjoy the air con, as an experiment to see what the new C# 7 stuff could help us with.

File: workers/unity/Packages/com.improbable.gdk.core/Logging/Logger.cs (hunk 1 above)
Comment (zeroZshadow) [reply workers/unity/Packages/com.improbable.gdk.core/Logging/Logger.cs]: I added this as an attempt to get rid of the pattern where every log would have to pass in the class it was logging for. I'm open for suggested alternatives.

Comment (MPeti01) [reply workers/unity/Packages/com.improbable.gdk.core/Logging/Logger.cs]: Why couldn't the wrapper also implement ILogDispatcher?

File: workers/unity/Assets/Playground/Scripts/Bootstrap.cs [hunk 3]
```diff
@@ -31,9 +28,9 @@ public void Awake()
             PlayerLoopManager.RegisterDomainUnload(DomainUnloadShutdown, 10000); // Clean up worlds and player loop
 
             Application.targetFrameRate = TargetFrameRate;
+#if UNITY_EDITOR
```
Comment (martinzlocha) [inline workers/unity/Assets/Playground/Scripts/Bootstrap.cs]: Doesn't this have to be inside of the if statement? Wont the code be invalid if UNITY_EDITOR is false?

File: workers/unity/Packages/com.improbable.gdk.core/Logging/ILogger.cs [hunk 4]
```diff
@@ -0,0 +1,9 @@
+using UnityEngine;
```
Comment (martinzlocha) [inline workers/unity/Packages/com.improbable.gdk.core/Logging/ILogger.cs]: Don't think we should have yet another logger interface. Instead we could consider making changes to ILogDispatcher

File: workers/unity/Packages/com.improbable.gdk.core/Logging/Logger.cs (hunk 1 above)
Comment (zeroZshadow) [reply workers/unity/Packages/com.improbable.gdk.core/Logging/Logger.cs]: Maybe the `ILogger` interface was a bit much, I could just make it a `Logger` directly? Its simply meant to be a tool to de-duplicate the class field passing.

File: workers/unity/Packages/com.improbable.gdk.core/Systems/CleanReactiveComponentsSystem.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (jamiebrynes7) [reply workers/unity/Packages/com.improbable.gdk.core/Systems/CleanReactiveComponentsSystem.cs]: I'm fine with this being `var` personally. I think C# tooling is sufficiently good that a healthy use of `var` is fine.

File: workers/unity/Packages/manifest.json [hunk 5]
```diff
@@ -5,10 +5,10 @@
     "com.improbable.gdk.playerlifecycle": "file:com.improbable.gdk.playerlifecycle",
     "com.improbable.gdk.testutils": "file:com.improbable.gdk.testutils",
     "com.improbable.gdk.transformsynchronization": "file:com.improbable.gdk.transformsynchronization",
-    "com.unity.burst": "0.2.4-preview.18",
+    "com.unity.burst": "0.2.4-preview.21",
```
Comment (jared-improbable) [inline workers/unity/Packages/manifest.json]: This could be split out into a separate PR.

**Comment by zeroZshadow** [comment]
Will redo this in separate bits when back from vacation
